File: app/src/main/java/com/sample/dhruvmj/waveapp/view/ProductListAdapter.java
package com.sample.dhruvmj.waveapp.view;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.sample.dhruvmj.waveapp.R;
import com.sample.dhruvmj.waveapp.databinding.ProductListItemViewBinding;
import com.sample.dhruvmj.waveapp.model.Product;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Dhruv
 */

public class ProductListAdapter extends RecyclerView.Adapter<ProductViewHolder> {
    private List<Product> productList;
    private Context context;

    ProductListAdapter(Context context) {
        this.context = context;
        productList = new ArrayList<>();
    }

    void setItems(List<Product> productList) {
        if (productList != null) {
            this.productList.clear();
            this.productList.addAll(productList);
            notifyDataSetChanged();
        }
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ProductListItemViewBinding binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.product_list_item_view, parent, false);
        return new ProductViewHolder(binding.getRoot());
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        if (position != RecyclerView.NO_POSITION) {
            Product product = productList.get(position);
            holder.productListItemViewBinding.setProduct(product);
        }
    }

    @Override
    public int getItemCount() {
        return productList.size();
    }
}

File: app/src/main/java/android/sample/wavechallenge/adapter/ProductListAdapter.kt
package android.sample.wavechallenge.adapter

import android.content.Context
import android.sample.wavechallenge.R
import android.sample.wavechallenge.model.Product
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.item_product_layout.view.*

class ProductListAdapter(
    private var context: Context?,
    private var itemList: List<Product>?
) : RecyclerView.Adapter<ProductListAdapter.ItemViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_product_layout, parent, false)
        return ItemViewHolder(view)
    }

    override fun getItemCount(): Int {
        return itemList?.size ?: 0
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val product = itemList?.get(position)
        product?.let {
            holder.tvName.text = product.name
            holder.tvPrice.text = context?.getString(R.string.formatted_price, product.price.toString())
        }
    }

    fun updateItems(products: List<Product>) {
        itemList = listOf()
        itemList = products
        notifyDataSetChanged()
    }

    class ItemViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var tvName: TextView = view.tv_name
        var tvPrice: TextView = view.tv_price
    }
}